File: bun-honojs/src/routes/params.ts
import { Hono } from "hono";
import { getCookie, setCookie } from "hono/cookie";
import { MAX_FILE_BYTES, NULL_BYTE, SNIFF_LEN, DEFAULT_LIMIT } from "../consts/defaults";
import {
  INVALID_JSON_BODY,
  INVALID_FORM_DATA,
  INVALID_MULTIPART,
  FILE_NOT_FOUND,
  FILE_SIZE_EXCEEDS,
  ONLY_TEXT_PLAIN,
  FILE_NOT_TEXT
} from "../consts/errors";

export const paramsRoutes = new Hono();

paramsRoutes.get("/search", (c) => {
  const q = c.req.query("q")?.trim() || "none";

  const limitStr = c.req.query("limit");
  const limitNum = Number(limitStr);
  const limit = Number.isSafeInteger(limitNum) ? limitNum : DEFAULT_LIMIT;

  return c.json({ search: q, limit });
});

paramsRoutes.get("/url/:dynamic", (c) => {
  const dynamic = c.req.param("dynamic");
  return c.json({ dynamic });
});

paramsRoutes.get("/header", (c) => {
  const header = c.req.header("X-Custom-Header")?.trim() || "none";
  return c.json({ header });
});

paramsRoutes.post("/body", async (c) => {
  let body: Record<string, unknown>;
  try {
    body = await c.req.json();
  } catch {
    return c.json({ error: INVALID_JSON_BODY }, 400);
  }

  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    return c.json({ error: INVALID_JSON_BODY }, 400);
  }

  return c.json({ body });
});

paramsRoutes.get("/cookie", (c) => {
  const cookie = getCookie(c, "foo")?.trim() || "none";
  setCookie(c, "bar", "12345", { maxAge: 10, httpOnly: true, path: "/" });
  return c.json({ cookie });
});

paramsRoutes.post("/form", async (c) => {
  const contentType = c.req.header("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("application/x-www-form-urlencoded") && !contentType.startsWith("multipart/form-data")) {
    return c.json({ error: INVALID_FORM_DATA }, 400);
  }

  let form: Record<string, string | File>;
  try {
    form = await c.req.parseBody();
  } catch {
    return c.json({ error: INVALID_FORM_DATA }, 400);
  }

  if (typeof form !== "object" || form === null || Array.isArray(form)) {
    return c.json({ error: INVALID_FORM_DATA }, 400);
  }

  const name = typeof form.name === "string" && form.name.trim() !== "" ? form.name.trim() : "none";

  const ageStr = typeof form.age === "string" && form.age.trim() !== "" ? form.age.trim() : "0";
  const ageNum = Number(ageStr);
  const age = Number.isSafeInteger(ageNum) ? ageNum : 0;

  return c.json({ name, age });
});

paramsRoutes.post("/file", async (c) => {
  const contentType = c.req.header("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("multipart/form-data")) {
    return c.json({ error: INVALID_MULTIPART }, 400);
  }

  let form: FormData;
  try {
    form = await c.req.formData();
  } catch {
    return c.json({ error: INVALID_MULTIPART }, 400);
  }

  const file = form.get("file");
  if (!(file instanceof File)) {
    return c.json({ error: FILE_NOT_FOUND }, 400);
  }
  if (!file.type || !file.type.startsWith("text/plain")) {
    return c.json({ error: ONLY_TEXT_PLAIN }, 415);
  }
  if (file.size > MAX_FILE_BYTES) {
    return c.json({ error: FILE_SIZE_EXCEEDS }, 413);
  }

  const buffer = await file.arrayBuffer();
  const data = new Uint8Array(buffer);
  if (data.length > MAX_FILE_BYTES) {
    return c.json({ error: FILE_SIZE_EXCEEDS }, 413);
  }

  const head = data.slice(0, SNIFF_LEN);
  if (head.includes(NULL_BYTE)) {
    return c.json({ error: FILE_NOT_TEXT }, 415);
  }

  if (data.includes(NULL_BYTE)) {
    return c.json({ error: FILE_NOT_TEXT }, 415);
  }

  let content: string;
  try {
    content = new TextDecoder("utf-8", { fatal: true }).decode(data);
  } catch {
    return c.json({ error: FILE_NOT_TEXT }, 415);
  }

  return c.json({
    filename: file.name,
    size: data.length,
    content
  });
});

File: http-servers/go/chi/internal/routes/params.go
package routes

import (
	"bufio"
	"cmp"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"slices"
	"strconv"
	"strings"
	"unicode/utf8"

	"chi-server/internal/consts"
	"chi-server/internal/utils"

	"github.com/go-chi/chi/v5"
)

func RegisterParams(r chi.Router) {
	r.Get("/search", handleSearchParams)
	r.Get("/url/{dynamic}", handleUrlParams)
	r.Get("/header", handleHeaderParams)
	r.Post("/body", handleBodyParams)
	r.Get("/cookie", handleCookieParams)
	r.Post("/form", handleFormParams)
	r.Post("/file", handleFileParams)
}

func handleSearchParams(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query()

	q := cmp.Or(strings.TrimSpace(query.Get("q")), "none")

	limit := consts.DefaultLimit
	limitStr := query.Get("limit")
	if limitStr != "" && !strings.Contains(limitStr, ".") {
		if n, err := strconv.ParseInt(limitStr, 10, 64); err == nil && n >= -(1<<53-1) && n <= (1<<53-1) {
			limit = int(n)
		}
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"search": q, "limit": limit})
}

func handleUrlParams(w http.ResponseWriter, r *http.Request) {
	dynamic := chi.URLParam(r, "dynamic")
	utils.WriteResponse(w, http.StatusOK, map[string]any{"dynamic": dynamic})
}

func handleHeaderParams(w http.ResponseWriter, r *http.Request) {
	header := cmp.Or(strings.TrimSpace(r.Header.Get("X-Custom-Header")), "none")
	utils.WriteResponse(w, http.StatusOK, map[string]any{"header": header})
}

func handleBodyParams(w http.ResponseWriter, r *http.Request) {
	defer func() { _ = r.Body.Close() }()

	var body map[string]any
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInvalidJSON, err.Error())
		return
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"body": body})
}

func handleCookieParams(w http.ResponseWriter, r *http.Request) {
	cookieVal, err := r.Cookie("foo")

	cookie := "none"
	if err == nil && cookieVal != nil {
		if trimmed := strings.TrimSpace(cookieVal.Value); trimmed != "" {
			cookie = trimmed
		}
	}

	http.SetCookie(w, &http.Cookie{Name: "bar", Value: "12345", MaxAge: 10, HttpOnly: true, Path: "/"})
	utils.WriteResponse(w, http.StatusOK, map[string]any{"cookie": cookie})
}

func handleFormParams(w http.ResponseWriter, r *http.Request) {
	defer func() { _ = r.Body.Close() }()

	contentType := strings.ToLower(r.Header.Get("Content-Type"))
	if !strings.HasPrefix(contentType, "application/x-www-form-urlencoded") && !strings.HasPrefix(contentType, "multipart/form-data") {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInvalidForm, consts.ErrExpectedFormContentType)
		return
	}

	if err := r.ParseForm(); err != nil {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInvalidForm, err.Error())
		return
	}

	name := cmp.Or(strings.TrimSpace(r.FormValue("name")), "none")

	age := 0
	ageStr := strings.TrimSpace(r.FormValue("age"))
	if ageStr != "" {
		if n, err := strconv.ParseInt(ageStr, 10, 64); err == nil && n >= -(1<<53-1) && n <= (1<<53-1) {
			age = int(n)
		}
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"name": name, "age": age})
}

func handleFileParams(w http.ResponseWriter, r *http.Request) {
	defer func() { _ = r.Body.Close() }()

	contentType := strings.ToLower(r.Header.Get("Content-Type"))
	if !strings.HasPrefix(contentType, "multipart/form-data") {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInvalidMultipart, consts.ErrExpectedMultipartContentType)
		return
	}

	if err := r.ParseMultipartForm(consts.MaxFileBytes); err != nil {
		if strings.Contains(strings.ToLower(err.Error()), "too large") {
			utils.WriteError(w, http.StatusRequestEntityTooLarge, consts.ErrFileSizeExceeded)
			return
		}
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInvalidMultipart, err.Error())
		return
	}

	file, fileHeader, err := r.FormFile("file")
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrFileNotFound, err.Error())
		return
	}
	defer func() { _ = file.Close() }()

	br := bufio.NewReader(file)

	head, err := br.Peek(consts.SniffLen)
	if err != nil && !errors.Is(err, io.EOF) {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInternal, err.Error())
		return
	}

	fileContentType := fileHeader.Header.Get("Content-Type")
	if fileContentType != "" {
		if !strings.HasPrefix(strings.ToLower(fileContentType), "text/plain") {
			utils.WriteError(w, http.StatusUnsupportedMediaType, consts.ErrInvalidFileType)
			return
		}
	} else {
		if mime := http.DetectContentType(head); !strings.HasPrefix(mime, "text/plain") {
			utils.WriteError(w, http.StatusUnsupportedMediaType, consts.ErrInvalidFileType)
			return
		}
	}

	if slices.Contains(head, consts.NullByte) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, consts.ErrNotPlainText)
		return
	}

	limited := io.LimitReader(br, consts.MaxFileBytes+1)
	data, err := io.ReadAll(limited)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, consts.ErrInternal, err.Error())
		return
	}
	if int64(len(data)) > consts.MaxFileBytes {
		utils.WriteError(w, http.StatusRequestEntityTooLarge, consts.ErrFileSizeExceeded)
		return
	}
	if slices.Contains(data, consts.NullByte) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, consts.ErrNotPlainText)
		return
	}
	if !utf8.Valid(data) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, consts.ErrNotPlainText)
		return
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{
		"filename": fileHeader.Filename,
		"size":     len(data),
		"content":  string(data),
	})
}
